### ISSUE TITLE
Investigate use of multiple useMappedState calls

### ISSUE THREAD

**Issue by ncphillips** [facebookincubator/redux-react-hook#12]
The following note is in the README:

> NOTE: You should only call useMappedState once per component, in the same way that you'd use connect from react-redux only once per component. Otherwise the component will have multiple store subscriptions and could rerender multiple times from a single store update.

I think this is a serious downside. One of the main benefits I would hope to get from using hooks with redux, is the ability to create a higher level abstraction. The purpose of doing this would be to hide the details of how my state is managed (i.e. that redux is used at all).

In the following example there are two custom hooks that use `redux-react-hook` under the hood. The actual component has no knowledge of Redux at all, it works entirely within its domain.

```javascript
//
// Individual components
//
import {useDispatch, useMappedState} from 'redux-react-hook';

export function DeleteButton({index}) {
  let todo = useTodo(index)
  let user = useUser()
  
  return (
    <button 
      disabled={user.permissions.deleteTodos && !todo.canDelete} 
      onClick={todo.deleteTodo}
    >
      Delete {todo.name}
    </button>
  );
}

function useTodo(index) {
  const mapState = useCallback(
    state => ({
      canDelete: state.todos[index].canDelete,
      name: state.todos[index].name,
    }),
    [index],
  );

  const {canDelete, name} = useMappedState(mapState);

  const dispatch = useDispatch();
  const deleteTodo = useCallback(
    () =>
      dispatch({
        type: 'delete todo',
        index,
      }),
    [index],
  );

  return { name, canDelete, deleteTodo }
}

function useUser() {
  const mapState = useCallback(
    state => ({ user: state.user }),
    [],
  );

  const { user } = useMappedState(mapState);

  return user
}
```

**Comment by ianobermiller** [comment]
To give some context from that note, my perspective here is working with an app where you have N connected items, where N is the thing the user is creating (think each item in a list is connected). In that case, switching from 1 subscription to 2 increases the number of subscriptions by N, which is not scalable. I doubt the extra subscriptions matter when you are doing this at a high level. In fact, the conjecture about an extra re-render might not even be true, I'd certainly need to test it.

That being said, I very much agree with you that the entire point of hooks is to be able to compose them and create custom hooks that do multiple things, including fetch data from the store. The first thing we should do here is assess if there is actually any non-negligible perf impact to subscribing multiple times in the same component (chiefly, does it cause multiple React commits). Second, we can look at alternate APIs -- perhaps it is even a separate library. For example, you could have a two step process:

```tsx
const storeSubscription = useRedux();
const val1 = useMappedState(storeSubscription, state => state.foo.val1);
const val2 = useMappedState(storeSubscription, state => state.foo.val2);
const user = useUser(storeSubscription);
const {name, canDelete, deleteTodo} = useTodo(storeSubscription, index);
```

You would have to pass the magic `storeSubscription` object to your custom hooks. Still not seamless, but since one hook cannot access component state created by another hook I'm not sure if we have a better option.

**Comment by ncphillips** [comment]
I renamed this issue to clarify that this is currently possible, but that the implications are not known at the moment. This is definitely an interesting problem, and maybe something about the Hooks API will change before it's released. 

The above API is not ideal but could work. Adding the store subscription as an optional argument could be a way to add this behaviour for performance sensitive areas, without breaking the API.

```typescript
type useMappedState<TState, TResult> = (
  mapState: (state: TState) => TResult,
  storeSubscription?: StoreSubscription
): TResult 
```

**Comment by ncphillips** [comment]
I was just reading through the effect of subscribing to the store. 

It looks like a `useMappedState` does not cause the component to re-render on each store update. Instead seems to only cause a re-render if the new `TResult` does not shallowEqual the old `TResult`. If this is the case, then I would only expect there to be an issue if both hooks accessed the same data. For example:

```
const user = useUser();
const userAgain = useUser();
```

Maybe a perf test would be helpful in determining if this is a problem or not.

**Comment by ianobermiller** [comment]
> It looks like a useMappedState does not cause the component to re-render on each store update

That is correct, it does a shallow comparison just like react-redux. There aren't any issues if you were to subscribe to the same data multiple times, because the previous mapped value that is capture is done for each subscription. It would be a little weird, however, to do a shallow comparison on something that isn't expected to be an object (you'd probably want to use an exact comparison instead).

I did a small test in this codesandbox, https://codesandbox.io/s/7w16rv9x91, see `TodoList.tsx`. In it, I call `useMappedState` twice (so it will do two store subscriptions. If you look at the console logs, however, the component is only rendered once for each store change. This is likely because redux is calling all the subscribers synchronously and React debounces multiple setState calls in the same frame. So at least from a component perf perspective this isn't an issue. The only thing would be to measure the impact of more store subscriptions on Redux, but I am inclined to say that performance-wise this won't be a huge deal.

**Comment by ncphillips** [comment]
That seems like a good doc update. Do you want to close this issue?

**Comment by ianobermiller** [comment]
Closed by b207d3c472f81be5ce3c6c4398cc1289aa2b54f6
